lib/Window.js

```diff
@@ -23,100 +23,165 @@ module.exports = (Item) => {
     }
 
     acceptClick (click) {
-      assert.ok(click.mouseButton === 0 || click.mouseButton === 1)
+      assert.ok(((click.slot >= 0 && click.slot < this.inventoryEnd) || click.slot === -999) && click.mouseButton >= 0 && click.mouseButton <= 6, 'invalid operation')
+
+      switch (click.mode) {
+        case 0:
+          assert.ok(click.mouseButton <= 2, 'invalid operation')
+          this.mouseClick(click)
+          break
+
+        case 1:
+          assert.ok(click.mouseButton <= 2, 'invalid operation')
+          this.shiftClick(click)
+          break
+
+        case 2:
+          assert.ok(click.mouseButton <= 8 && !this.selectedItem, 'invalid operation')
+          this.numberClick(click)
+          break
+
+        case 4:
+          assert.ok(click.mouseButton < 2 && !this.selectedItem, 'invalid operation')
+          this.dropClick(click)
+          break
+
+        default:
+          assert.ok(false, 'unimplemented')
+      }
+    }
+
+    mouseClick (click) {
       if (click.slot === -999) {
-        this.acceptOutsideWindowClick(click)
-      } else if (click.slot >= this.inventoryStart && click.slot < this.inventoryEnd) {
-        this.acceptInventoryClick(click)
-      } else if (click.slot === this.craftingResultSlot) {
-        this.acceptCraftingClick(click)
+        this.dropSelectedItem(!click.mouseButton)
```

Comment:
I'd recommend replacing ``!click.mouseButton`` with ``click.mouseButton === 0`` (also line 59) as it's much more understandable :)